packages/backend-common/src/database/DatabaseManager.ts

```diff
@@ -111,6 +111,30 @@ export class DatabaseManager {
     return { getClient, migrations };
   }
 
+  /**
+   * Drops database for given plugin.
+   *
+   * @param pluginId - Plugin to drop the database for.
+   */
+  async dropDatabase(
+    pluginId: string,
+    deps?: {
+      lifecycle: LifecycleService;
+      pluginMetadata: PluginMetadataService;
+    },
+  ) {
+    const pluginConfig = new ConfigReader(
+      this.getConfigForPlugin(pluginId) as JsonObject,
+    );
+    const client = createDatabaseClient(pluginConfig, {}, deps);
```

Comment:
Hm, the client that's constructed when _creating_ databases is unique, database specific, and has higher privileges; [see the code here](https://github.com/backstage/backstage/blob/bb6bce457ed069a38ef8d16ef38602972c7735c5/packages/backend-common/src/database/connectors/postgres.ts#L125) and note that that's in the postgres connector, not generic code. The one [in the mysql connector](https://github.com/backstage/backstage/blob/bb6bce457ed069a38ef8d16ef38602972c7735c5/packages/backend-common/src/database/connectors/mysql.ts#L150) is, again, different. You would need to add a dropDatabase and dropSchema into the connectors system for this (where only the pg connector actually implements the dropSchema).

Hm. Let me back up a bit.

All of this makes me ponder if it's unfortunate that the TestDatabases use this class at all. Deleting things just isn't part of what it was meant to do, and it kind of shows at this point. TestDatabases already knows that the BASE credentials it has been given are enough to do admin actions, because of how it's designed; it doesn't need to elevate itself. DatabaseManager which is in the production path does not have that luxury. Now the proposed code in here "leaks" that difference into the DatabaseManager class. The TestDatabases don't care at all about plugin division modes either - that's only relevant in production.

OK, in summary, I think this class should actually be left unchanged. Is there some way of ONLY putting the deletion code inside TestDatabases? Then we can as a separate task in the future see if we can just plain detach TestDatabases from DatabaseManager.

Reply:
Yup, I totally agree with that. I had hard time understanding the responsibility of the database manager, so thanks for clarifying that. I'll make the proposed changes later.

Reply:
Done.